src/Akeneo/Category/back/Infrastructure/Storage/Sql/GetCategoryTreesSql.php

```diff
@@ -0,0 +1,110 @@
+        $condition['types'] = ['ids' => Connection::PARAM_INT_ARRAY];
+
+        return $this->execute($condition);
+    }
+
+    private function execute(array $condition = []): ?array
+    {
+        $sqlAnd = $condition['sqlAnd'] ?? '';
+        $sqlParams = $condition['params'] ?? [];
+        $sqlTypes = $condition['types'] ?? [];
+
+        $sqlQuery = <<<SQL
+            WITH category_tree_translation as (
+                SELECT category.code, JSON_OBJECTAGG(translation.locale, translation.label) as translations
+                FROM pim_catalog_category category
+                JOIN pim_catalog_category_translation translation ON translation.foreign_key = category.id
+                WHERE category.parent_id IS NULL 
+                AND category.root = category.id
+                $sqlAnd
+                GROUP BY category.code
+            )
+            SELECT
+                category.id,
+                category.code,
+                category_tree_translation.translations,
+                BIN_TO_UUID(template.uuid) as template_uuid,
+                template.labels as template_labels
+            FROM 
+                pim_catalog_category category
+                LEFT JOIN category_tree_translation ON category_tree_translation.code = category.code
+                LEFT JOIN pim_catalog_category_tree_template ctt ON ctt.category_tree_id = category.id
+                LEFT JOIN pim_catalog_category_template template ON template.uuid = ctt.category_template_uuid
+            WHERE category.parent_id IS NULL 
+            AND category.root = category.id
+            $sqlAnd
+            ORDER BY category.created DESC
+        SQL;
+
+        $results = $this->connection->executeQuery(
+            $sqlQuery,
+            $sqlParams,
+            $sqlTypes
+        )->fetchAllAssociative();
+
+        if (empty($results)) {
+            return null;
+        }
+
+        return array_map(function ($result) {
+            return new CategoryTree(
+                new CategoryId((int)$result['id']),
+                new Code($result['code']),
+                $result['translations'] ?
+                    LabelCollection::fromArray(
+                        json_decode(
+                            $result['translations'],
+                            true,
+                            512,
+                            JSON_THROW_ON_ERROR
+                        )
+                    ) : null,
+                $result['template_uuid'] ? TemplateUuid::fromString($result['template_uuid']) : null,
+                $result['template_labels'] ?
+                    LabelCollection::fromArray(
+                        json_decode(
+                            $result['template_labels'],
+                            true,
+                            512,
+                            JSON_THROW_ON_ERROR
+                        )
+                    ) : null,
+            );
+        }, $results);
```

Comment:
The creation of a CategoryTree object could be defined inside the model itself by using a named constructor.
The code would be more readable and the "complexity" of the creation  (validation of the data, use of value objects ...) will be owned by the model in one place. What do you think?

The name of the constructor could be ` CategoryTree::fromNormalzed` or ` CategoryTree::fromArray` for instance.

```suggestion
        return array_map(function ($result) {
            return CategoryTree::fromNormalzed(
               (int) $result['id'],
                $result['code'],
                $result['translations'] ?? null
                $result['template_uuid'] ?? null,
                $result['template_labels'] ?? null,
            );
        }, $results);
```

Reply:
You are right, I'll do it